datasets/utils/secret.py

```diff
@@ -0,0 +1,210 @@
+            if key is None: entire secret
+        """
+        self._variable_validation()
+
+        if self.cluster_secret_name:
+            return self._fetch_cluster_secret()
+        if self.aws_secret_arn:
+            return self._fetch_aws_secret()
+        if self.env_var:
+            return self._fetch_env_secret()
+        if self.raw_secret:
+            return self._fetch_raw_secret()
+
+    def _variable_validation(self):
+        """
+        Ensure exact one secret source variable provided.
+        Ensure key is of str type if provided
+        """
+        secret_source_variables = ["cluster_secret_name", "aws_secret_arn", "env_var", "raw_secret"]
+
+        secret_source_variable_used = [
+            getattr(self, variable) is not None for variable in secret_source_variables
+        ]
+
+        if secret_source_variable_used.count(True) != 1:
+            raise ValueError("Must provide exact one secret source variable!")
+
+        if self.key is not None and not isinstance(self.key, str):
+            raise ValueError("key should be an string!")
+
+    def _fetch_cluster_secret(self) -> secret_return_type:
+        # Try to fetch from cache first
+        global secret_cache
+        secret_from_cache = secret_cache["cluster_secret"].get(self.cluster_secret_name)
+        if secret_from_cache is not None and not self.force_reload:
+            logger.info(f"Using secret from cache {self.cluster_secret_name}")
+            secret_value = secret_from_cache
+        else:
+            config.load_incluster_config()
+            core_api = client.CoreV1Api()
+            namespace = get_current_namespace()
+            for attempt in Retrying(stop=stop_after_attempt(3), wait=wait_fixed(5), reraise=True):
+                with attempt:
+                    raw_secret_value = core_api.read_namespaced_secret(
+                        self.cluster_secret_name, namespace
+                    ).data
+
+            # Ensure we convert to Dict[str, str] type
+            secret_value = {}
+            for k, v in raw_secret_value.items():
+                # The last decode("utf-8") converts bytes type to str
+                secret_value[str(k)] = base64.b64decode(v).decode("utf-8")
+            secret_cache["cluster_secret"][self.cluster_secret_name] = secret_value
+
+        return secret_value[self.key] if self.key is not None else secret_value
+
+    def _fetch_aws_secret(self) -> secret_return_type:
+        global secret_cache
+        secret_from_cache = secret_cache["aws_secret"].get(self.aws_secret_arn)
+        if secret_from_cache is not None and not self.force_reload:
+            logger.info(f"Using secret from cache {self.aws_secret_arn}")
+            secret_value = secret_from_cache
+        else:
+            # We have AWS_REGION env var injected in most of our workloads
+            secrets_manager_client = boto3.Session(region_name=os.getenv("AWS_REGION", "us-west-2")).client(
```

Comment:
Yeah generally because of that ParameterStore way is already covered by the k8s namespaced secret option, and this secret manager option is to cover the remaining/legacy resources that many of our customers do still have lots of secretmanager use cases.

We can always add additional e.g. `aws_parameter_arn` option if there's necessity to support in the future.

Reply:
I'd say we call this method `secretsmanager` instead (ie `fetch_aws_secretsmanager` or something like that) of `aws_secret` then. I agree with your assessment that `ParameterStore` largely should be covered by K8s secrets given that's the ZGCP paved path. I do think being more explicit in naming will make it easier to make room in the future if we need to support other AWS services.